Support callback with a new port

Br4nd0R opened the pull request:
# Changes
* CumulusCI uses a new port (7788) for the built-in connected app to lessen the chances that the port is in use. 
* CumulusCI now checks if the port associated with a callback URL/redirect URI is in use during OAuth2 flows, and if so, raises a more friendly error.

Br4nd0R commented on cumulusci/oauth/tests/test_client.py:
```diff
@@ -75,13 +76,26 @@ def httpd_thread(tester_class, oauth_client):
+        yield oauth_client
     finally:
         oauth_client.httpd.shutdown()
+        oauth_client.httpd.server_close()
```
This is the line that fixed the issue 🤷‍♂️

Br4nd0R commented:
@davisagli - So I _think_ I'm seeing the same odd behavior in all environments and not sure how to get around this. The current `test_create_httpd__port_already_in_use` works for osx/linux. To simulate a port already being used I use the `httpd_thread()` context manager to start a http daemon thread on port 7788. Then I simply calll `client._create_httpd()` which attempts to create another daemon process on the same port, and viola, we get the error we want. This all makes sense to me. 

But if instead of using the context manager I try to simulate the port being in use by binding a socket to it like so:
```
    def test_create_httpd__port_already_in_use(self, client):
        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        sock.bind(("localhost", 7788))
        with pytest.raises(OAuth2Error, match=PORT_IN_USE_ERR):
            client._create_httpd()
        sock.close()
```
Then it will fail on the call to`sock.bind()`. But what is occupying that socket?!
I believe this is caused by the instance of `HTTPServer` used by the other tests, because if you move this test up to the top of the class, then the same code will pass.

The current way the test is written does not cause the error to occur on windows, and the only other way I know how to repro this error is by direct binding of a socket (which causes the above error), I'm a little stuck unless I can figure out the issue with `socket.bind()` and why it's not playing nicely with the `HTTPServer`.

davisagli reviewed (commented):
@Br4nd0R I did some experimenting and am not sure why the test that works in OS X is not working on Windows (why doesn't it throw an error?)  Before we spend time trying to come up with another test that may or may not simulate real behavior on Windows, can we check what is currently actually happening on Windows with this branch? i.e. if you run cci org connect in one terminal and then run it again from another terminal without stopping the first one, does it work or error or what?

Br4nd0R commented on cumulusci/oauth/client.py:
```diff
@@ -40,6 +40,8 @@
 )
 PROD_LOGIN_URL = os.environ.get("SF_PROD_LOGIN_URL") or "https://login.salesforce.com"
+PORT_IN_USE_ERR = "Cannot listen for callback, as port {} is already in use."
+GENERIC_OAUTH_ERROR = "An error occurred during the OAuth process: {}"
```
`PORT_IN_USE_ERR` is being used, but `GENERIC_OAUTH_ERROR` is not.

Br4nd0R commented:
@davisagli  - Current behavior in windows is that both commands open a web browser to the expected login page without error.

Also, switching the test to use `socket.bind()` fails in _both_ osx/windows contexts. It's just that for some reason the `httpd_thread()` context manager in `test_client.py` seems to only replicate the issue in non-windows based environments. This seems to confirm what you were suggesting above: that this is really a fix for an issue in non-windows enviornments.

I've gone ahead and fixed the things you noted above, and also moved the opening the tab in the webbrowser to _after_ we start the http daemon process so that if an error is encountered we don't open a new tab for the user.
